maskit/ensembles.py

```diff
@@ -0,0 +1,179 @@
+from collections import deque
+from typing import Dict, NamedTuple, Optional
+from pennylane import numpy as np
+
+from maskit.masks import MaskedCircuit
+
+
+class EnsembleResult(NamedTuple):
+    branch: MaskedCircuit
+    """branch that performs best"""
+    branch_name: str
+    """name of branch as configured"""
+    cost: float
+    """cost of selected branch"""
+    gradient: np.ndarray
+    """gradient of selected branch"""
+    brutto_steps: int
+    """training steps including all other branches"""
+    netto_steps: int
+    """training steps for selected branch"""
+    brutto: int
+    """training wrt count of parameters including all other branches"""
+    netto: int
+    """training wrt count of paramemters for selected branch"""
+    ensemble: bool
+    """True in case the ensemble was evaluated, otherwise False"""
+
+
+class Ensemble(object):
+    __slots__ = ("dropout", "perturb")
+
+    def __init__(self, dropout: Optional[Dict], *args, **kwargs):
+        super().__init__()
+        self.dropout = dropout
+        self.perturb = True
+
+    def _branch(
+        self, masked_circuit: MaskedCircuit
+    ) -> Optional[Dict[str, MaskedCircuit]]:
+        if not self.perturb or self.dropout is None:
+            return None
+        branches = {}
+        for key in self.dropout:
+            branches[key] = MaskedCircuit.execute(masked_circuit, self.dropout[key])
+        return branches
+
+    def step(
+        self, masked_circuit: MaskedCircuit, optimizer, *args, ensemble_steps: int = 0
```

Comment:
Any reason why ``args[0]`` is not a named-positional parameter? It is expected throughout the method several times.

Reply:
No special reason, great point :)